File: reference/python-nlp-service/skills.py
"""Skill dictionary shared by resume and job-description parsing."""

SKILLS: dict[str, dict] = {
    "Python": {"category": "Programming", "aliases": ["python3"]},
    "Java": {"category": "Programming", "aliases": []},
    "JavaScript": {"category": "Programming", "aliases": ["js", "es6"]},
    "TypeScript": {"category": "Programming", "aliases": ["ts"]},
    "C++": {"category": "Programming", "aliases": ["cpp"]},
    "SQL": {"category": "Data", "aliases": []},
    "MySQL": {"category": "Data", "aliases": []},
    "PostgreSQL": {"category": "Data", "aliases": ["postgres"]},
    "MongoDB": {"category": "Data", "aliases": ["mongo"]},
    "Pandas": {"category": "Data", "aliases": []},
    "NumPy": {"category": "Data", "aliases": ["numpy"]},
    "Excel": {"category": "Data", "aliases": ["ms excel"]},
    "Power BI": {"category": "Data", "aliases": ["powerbi"]},
    "Tableau": {"category": "Data", "aliases": []},
    "Statistics": {"category": "Data", "aliases": ["statistical analysis"]},
    "A/B Testing": {"category": "Data", "aliases": ["ab testing", "split testing"]},
    "Machine Learning": {"category": "AI", "aliases": ["ml"]},
    "Deep Learning": {"category": "AI", "aliases": []},
    "NLP": {"category": "AI", "aliases": ["natural language processing"]},
    "Scikit-learn": {"category": "AI", "aliases": ["sklearn", "scikit learn"]},
    "TensorFlow": {"category": "AI", "aliases": []},
    "PyTorch": {"category": "AI", "aliases": []},
    "React": {"category": "Web", "aliases": ["react.js", "reactjs"]},
    "Node.js": {"category": "Web", "aliases": ["node", "nodejs"]},
    "Express.js": {"category": "Web", "aliases": ["express"]},
    "REST APIs": {"category": "Web", "aliases": ["rest api", "restful"]},
    "HTML5": {"category": "Web", "aliases": ["html"]},
    "CSS3": {"category": "Web", "aliases": ["css"]},
    "Flask": {"category": "Web", "aliases": []},
    "Django": {"category": "Web", "aliases": []},
    "Docker": {"category": "Tools", "aliases": []},
    "Git": {"category": "Tools", "aliases": []},
    "GitHub": {"category": "Tools", "aliases": []},
    "Linux": {"category": "Tools", "aliases": []},
    "AWS": {"category": "Cloud", "aliases": ["amazon web services"]},
    "Azure": {"category": "Cloud", "aliases": []},
    "Jupyter": {"category": "Tools", "aliases": ["jupyter notebook"]},
    "VS Code": {"category": "Tools", "aliases": ["visual studio code"]},
    "Communication": {"category": "Soft skills", "aliases": ["communication skills"]},
    "Teamwork": {"category": "Soft skills", "aliases": ["collaboration"]},
    "Problem Solving": {"category": "Soft skills", "aliases": ["problem-solving"]},
}


def all_terms() -> list[tuple[str, str, str]]:
    """Return (canonical_skill, category, searchable_term) triples."""
    out: list[tuple[str, str, str]] = []
    for name, meta in SKILLS.items():
        out.append((name, meta["category"], name.lower()))
        for alias in meta["aliases"]:
            out.append((name, meta["category"], alias.lower()))
    return out

File: src/lib/nlp/skills.ts
/**
 * Skill dictionary.
 *
 * Every entry is a canonical skill name plus optional aliases that are also
 * accepted while scanning text. Extend this file to teach the analyzer new
 * skills; no other module needs to change.
 */

export interface SkillDefinition {
  name: string;
  category: string;
  aliases?: string[];
}

const DICTIONARY: Record<string, Array<[string, string[]?]>> = {
  Programming: [
    ["Python"],
    ["Java", ["core java"]],
    ["C++", ["cpp", "c plus plus"]],
    ["C"],
    ["JavaScript", ["js", "es6"]],
    ["TypeScript", ["ts"]],
    ["Go", ["golang"]],
    ["R"],
  ],
  Data: [
    ["SQL"],
    ["MySQL"],
    ["PostgreSQL", ["postgres"]],
    ["MongoDB", ["mongo"]],
    ["Pandas"],
    ["NumPy", ["numpy"]],
    ["ETL"],
    ["Data Cleaning"],
    ["Data Visualization"],
  ],
  Analytics: [
    ["Power BI", ["powerbi"]],
    ["Excel", ["ms excel", "advanced excel"]],
    ["Tableau"],
    ["Statistics", ["statistical analysis"]],
    ["A/B Testing", ["ab testing"]],
  ],
  "Machine Learning": [
    ["Machine Learning", ["ml"]],
    ["Deep Learning"],
    ["TensorFlow"],
    ["Scikit-learn", ["sklearn", "scikit learn"]],
    ["Keras"],
    ["PyTorch"],
    ["NLP", ["natural language processing"]],
    ["Model Deployment"],
  ],
  "Computer Vision": [
    ["OpenCV", ["open cv"]],
    ["Object Detection", ["yolo"]],
    ["Image Classification"],
    ["Image Processing"],
  ],
  Web: [
    ["React", ["react.js", "reactjs"]],
    ["Node.js", ["nodejs", "node js"]],
    ["Express.js", ["expressjs", "express"]],
    ["HTML", ["html5"]],
    ["CSS", ["css3"]],
    ["REST API", ["rest apis", "restful api", "rest"]],
    ["Flask"],
    ["FastAPI", ["fast api"]],
    ["Django"],
  ],
  Tools: [
    ["Git"],
    ["GitHub"],
    ["Jupyter", ["jupyter notebook"]],
    ["VS Code", ["visual studio code"]],
    ["Docker"],
    ["Linux"],
    ["Postman"],
    ["AWS"],
    ["Azure"],
  ],
  "Soft Skills": [
    ["Communication"],
    ["Teamwork"],
    ["Problem Solving"],
    ["Leadership"],
    ["Time Management"],
  ],
};

export const SKILL_DICTIONARY: SkillDefinition[] = Object.entries(DICTIONARY).flatMap(
  ([category, entries]) =>
    entries.map(([name, aliases]) => ({
      name,
      category,
      ...(aliases ? { aliases } : {}),
    })),
);

export const SKILL_CATEGORIES = Object.keys(DICTIONARY);

export function findSkillDefinition(name: string): SkillDefinition | undefined {
  const lowered = name.toLowerCase();
  return SKILL_DICTIONARY.find((skill) => skill.name.toLowerCase() === lowered);
}
